airflow/migrations/versions/852ae6c715af_add_templated_fields_rendered_table.py

```diff
@@ -0,0 +1,103 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Add TemplatedFieldsRendered table
+
+Revision ID: 852ae6c715af
+Revises: 1d3a23a719bf
+Create Date: 2019-12-10 22:19:18.034961
+
+"""
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import mysql
+
+# revision identifiers, used by Alembic.
+revision = '852ae6c715af'
+down_revision = '1d3a23a719bf'
+branch_labels = None
+depends_on = None
+
+TABLE_NAME = 'templated_fields_rendered'
+INDEX_NAME = 'idx_tfr_dag_task_date'
+
+
+def upgrade():
+    """Apply Add TemplatedFieldsRendered table"""
+    json_type = sa.JSON
+    conn = op.get_bind()  # pylint: disable=no-member
+
+    if conn.dialect.name != "postgresql":
+        # Mysql 5.7+/MariaDB 10.2.3 has JSON support. Rather than checking for
+        # versions, check for the function existing.
+        try:
+            conn.execute("SELECT JSON_VALID(1)").fetchone()
+        except sa.exc.OperationalError:
+            json_type = sa.Text
+
+    op.create_table(
+        TABLE_NAME,  # pylint: disable=no-member
+        sa.Column('dag_id', sa.String(length=250), nullable=False),
+        sa.Column('task_id', sa.String(length=250), nullable=False),
+        sa.Column('execution_date', sa.DateTime(), nullable=False),
+        sa.Column('rendered_fields', json_type(), nullable=True),
+        sa.PrimaryKeyConstraint('dag_id', 'task_id', 'execution_date')
+    )
+    op.create_index(  # pylint: disable=no-member
+        INDEX_NAME,
+        TABLE_NAME,
+        ['dag_id', 'task_id', 'execution_date'],
+    )
+
+    if conn.dialect.name == "mysql":
+        conn.execute("SET time_zone = '+00:00'")
```

Comment:
We've got this in earlier migrations - I don't think we need it here too do we? Also: how about we just don't use a "TIMESTAMP" type, but "DATETIME" instead?

Reply:
It won't contain the timezones though if we just use DATETIME! It might have undesirable effects when comparing current time vs this time!

Reply:
Really? https://www.eversql.com/mysql-datetime-vs-timestamp-column-types-which-one-i-should-use/ would imply otherwise.

(Also TIMESTAMP only supports dates up to 2038.)

Reply:
From the same [article](https://www.eversql.com/mysql-datetime-vs-timestamp-column-types-which-one-i-should-use/) :

>To summarize, if you want to serve your date and time data the same way regardless of timezones, you can use the DATETIME type (which will also allow you to use all the date & type functions built in MySQL). Otherwise, you can use TIMESTAMP and serve the data on a per-timezone basis.

Also, from https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.DateTime
>For timezone support, use at least the TIMESTAMP datatype, if not the dialect-specific datatype object.

Reply:
I read that and my conclusion is we should use DATETIME for mysql - Airflow puts in checks to ensure that we only ever store UTC datetimes in the DB.

By using a datetime column the migrations become a lot simpler.